File: go/secret-handshake/secret_handshake.go
package secret

// Handshake returns an array of strings based on the binary representation of a number
func Handshake(num uint) []string {
	var res []string
	if num&1 == 1 {
		res = append(res, "wink")
	}
	if num&2 == 2 {
		res = append(res, "double blink")
	}
	if num&4 == 4 {
		res = append(res, "close your eyes")
	}
	if num&8 == 8 {
		res = append(res, "jump")
	}
	if num&16 == 16 {
		for i, j := 0, len(res)-1; i < j; i, j = i+1, j-1 {
			res[i], res[j] = res[j], res[i]
		}
	}
	return res
}

File: python/secret-handshake/secret_handshake.py
codes = {
  'wink': 1,
  'double blink': 2,
  'close your eyes': 4,
  'jump': 8,
}

actions = {
  1: 'wink',
  2: 'double blink',
  4: 'close your eyes',
  8: 'jump',
}

def handshake(code):
  res = [actions[c] for c in [1, 2, 4, 8] if code & c]
  if code & 16:
    res.reverse()
  return res

def secret_code(actions):
  sum = 0
  is_reversed = False
  for i, action in enumerate(actions):
    if not is_reversed and i != len(actions) - 1 and codes[actions[i]] > codes[actions[i + 1]]:
      sum += 16
      is_reversed = True
    sum += codes[action] 
  return sum